docs/smt.md

```diff
@@ -0,0 +1,225 @@
+maker. For the purposes of this example we focus only on the core logic, and ignore e.g.
+interactions with the underlying tokens.
+
+The `Amm` has two state variables representing it's underlying token blances, `reserve0` and
+`reserve1`. These are initialized to `1000` units each upon creation.
+
+We define a single invariant, that the product of the two reserves should never decrease. This is an
+important safety property, if it is violated, an attacker will be able to execute a sequence of
+trades that can drain all funds from the exchange.
+
+```act
+behaviour init of Amm
+interface constructor()
+
+creates
+
+    uint256 reserve0 := 1000
+    uint256 reserve1 := 1000
+
+invariants
+
+    1000 * 1000 <= reserve0 * reserve1
+```
+
+Next, we define the two `swap` methods, that allow a user to exchange tokens, these methods take an
+amount (`amt`) of input reserves, and the `Amm` will adjust the size of the other reserve according
+to the `x * y == k` constant product formula.
+
+```act
+behaviour swap0 of Amm
+interface swap0(uint256 amt)
+
+iff in range uint256
+
+    reserve0 + amt
+
+storage
+
+    reserve0 => reserve0 + amt
+    reserve1 => (reserve0 * reserve1) / (reserve0 + amt)
+```
+
+```act
+behaviour swap1 of Amm
+interface swap1(uint256 amt)
+
+iff in range uint256
+
+    reserve1 + amt
+
+storage
+
+    reserve0 => (reserve0 * reserve1) / (reserve1 + amt)
+    reserve1 => reserve1 + amt
+```
+
+If we run `act prove` against the above spec, we find that there is in fact an error. If the `Amm`
+starts with `reserve0 == 1` and `reserve1 == 1,000,000`, and `swap0` is called with `amt == 2`, the
+final state of the `Amm` will be `reserve0 == 3` and `reserve1 == 333,333`. In this case `3 *
+333,333` is `999,999`, and in fact the product of the reserves has decreased slightly due to
+imprecision introduced by the EVM's flooring division.
+
+A safe specification for the `swap` methods is as follows. Notice the extra `+1` added to the output
+reserve in both cases. With this implementation, the rounding error is now in favor of the pool
+instead of the trader, and the contract is now safe against this particular attack.
+
+If we again run `act prove` against the fixed specification, we see that the invariant holds for all
+possible executions of the contract.
+
+
+```act
+behaviour swap0 of Amm
+interface swap0(uint256 amt)
+
+iff in range uint256
+
+    reserve0 + amt
+
+storage
+
+    reserve0 => reserve0 + amt
+    reserve1 => (reserve0 * reserve1) / (reserve0 + amt) + 1
+```
+
+```act
+behaviour swap1 of Amm
+interface swap1(uint256 amt)
+
+iff in range uint256
+
+    reserve1 + amt
+
+storage
+
+    reserve0 => (reserve0 * reserve1) / (reserve1 + amt) + 1
+    reserve1 => reserve1 + amt
+```
+
+### Underpowered Invariant
+
+Due to the inductive nature of the proof, there are some true invariants that the SMT backend is
+unable to prove. For example, consider the following state machine:
```

Comment:
I'm not sure about this part. It's not like the SMT solver wouldn't be able to prove that `x < 9` is an invariant, the backend is just not interested in non-inductive invariants.

Reply:
To clarify: I think the example is good to distinguish inductive vs non-inductive invariants, I'm just not sure about the wording above.

Reply:
Any suggestions for better phrasing? I already have "due to the inductive nature of the proof" in the sentance above?